core/src/main/java/com/linecorp/armeria/client/HttpChannelPool.java

```diff
@@ -373,24 +393,43 @@ private void connect(SessionProtocol desiredProtocol, PoolKey key, ChannelAcquis
      * A low-level operation that triggers a new connection attempt. Used only by:
      * <ul>
      *   <li>{@link #connect(SessionProtocol, PoolKey, ChannelAcquisitionFuture,
-     *       ClientConnectionTimingsBuilder)} - The pool has been exhausted.</li>
+     *       ClientConnectionTimingsBuilder, ProxyContext)} - The pool has been exhausted.</li>
      *   <li>{@link HttpSessionHandler} - HTTP/2 upgrade has failed.</li>
      * </ul>
      */
     void connect(SocketAddress remoteAddress, SessionProtocol desiredProtocol,
-                 Promise<Channel> sessionPromise) {
+                 ProxyContext proxyContext, Promise<Channel> sessionPromise) {
+
         final Bootstrap bootstrap = getBootstrap(desiredProtocol);
-        final ChannelFuture connectFuture = bootstrap.connect(remoteAddress);
+
+        final Channel channel = bootstrap.register().channel();
+        channel.attr(PROXY_CONTEXT).set(proxyContext);
```

Comment:
Question: I thought `ProxyContext` seems to be a `request-scoped` object, not a `channel-scoped` object. Because `ProxyContext` contains a request URI. 🧐

Reply:
Yeah, `ProxyConfig` is connection-scoped though.

Reply:
🤔 
I agree the current code  may cause confusion.
(As @ikhoon pointed out, `reqUri` is set for each connection indicating the `URI` used when opening the connection. This `uri` may have no relation with other requests that may use this connection, potentially leading to bugs or unexpected behavior)

We can't avoid using URI since `ProxySelector` API highly depends on it.
Instead, I've tried to generify `reqUri` (by only leaving the host/port) and renamed it to `connUri` (or `connectionUri`).

Additionally, it seemed like `PoolKey` contains all related information needed (`host`, `port`), so I've modified `channel.attr` to remember `PoolKey` instead. (and removed `ProxyContext`)

Let me know if this looks better than the previous impl.

Reply:
Alternatively, we may also choose to revert back to the previous impl., or explore abandoning compatibility with `ProxySelector` and see how it looks.

Reply:
That's a good point, @jrhee17. I'm not sure `ProxySelector` will ever need to look into other than `(SessionProtocol, authority (or host and port))`. How about this:

- Make `ProxyConfigSelector` require `SessionProtocol` and `Endpoint` only.
- When calling `ProxySelector`, create a URI whose path is `/`, e.g. `http://foo.com/`